### PR TITLE
Support beforeSendTransaction

### LINKED ISSUE #1151: Support beforeSendTransaction
### Description

https://github.com/getsentry/sentry-docs/pull/5814

We need to create a `beforeSendTransaction` callback.

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

Closes #1151

## :green_heart: How did you test it?

Added unit tests

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps

@marandaneto There seem to be some oddities with serialization/deserialization with event properties when the instance is a transaction. Maybe it's just an issue in my test setup.

### CODE REVIEW

File: dart/lib/src/sentry_options.dart [hunk 1]
```diff
@@ -423,6 +427,12 @@ typedef BeforeSendCallback = FutureOr<SentryEvent?> Function(
   Hint? hint,
 });
 
+/// This function is called with an SDK specific transaction object and can return a modified transaction
+/// object or nothing to skip reporting the transaction
+typedef BeforeSendTransactionCallback = FutureOr<SentryTransaction?> Function(
+  SentryTransaction transaction,
+);
```
Comment (marandaneto) [inline dart/lib/src/sentry_options.dart:436]:
```suggestion
typedef BeforeSendTransactionCallback = FutureOr<SentryTransaction?> Function(
  SentryTransaction transaction,
  Hint? hint,
);
```

File: dart/lib/src/sentry_client.dart [hunk 2]
```diff
@@ -366,7 +361,44 @@ class SentryClient {
...
+    SentryEvent? eventOrTransaction = event;
+
+    final beforeSend = _options.beforeSend;
+    final beforeSendTransaction = _options.beforeSendTransaction;
+    String beforeSendName = 'beforeSend';
+
+    try {
+      if (event is SentryTransaction && beforeSendTransaction != null) {
+        beforeSendName = 'beforeSendTransaction';
+        eventOrTransaction = await beforeSendTransaction(event);
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart:377]: Pass the hint.

File: dart/lib/src/sentry_client.dart [hunk 3]
```diff
@@ -366,7 +361,44 @@ class SentryClient {
...
+      _options.logger(
+        SentryLevel.error,
+        'The $beforeSendName callback threw an exception',
+        exception: exception,
+        stackTrace: stackTrace,
+      );
+    }
+
+    if (eventOrTransaction == null) {
+      _recordLostEvent(event, DiscardReason.beforeSend);
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart:391]: We need to check if there's a discard reason specifically for the transaction.

File: dart/test/sentry_client_test.dart [hunk 4]
```diff
@@ -1458,6 +1538,21 @@ FutureOr<SentryEvent?> beforeSendCallback(SentryEvent event, {Hint? hint}) {
     ..sdk!.addPackage('test-pkg', '1.0');
 }
 
+FutureOr<SentryTransaction?> beforeSendTransactionCallback(
+    SentryTransaction transaction) {
+  return transaction
+    ..tags!.addAll({'theme': 'material'})
+    ..extra!['host'] = '0.0.0.1';
+
+  // TODO Check why event properties are dropped (constructor/copy). Is this expected?
```
Comment (marandaneto) [inline dart/test/sentry_client_test.dart]: Could be a bug, has to be checked, did you try to use `copyWith` instead? some getters return an immutable list or make it immutable in the ctor.

File: dart/lib/src/sentry_options.dart (hunk 1 above)
Comment (denrase) [reply dart/lib/src/sentry_options.dart:436]: The `captureException` method of the sentry client does not have a hint. https://github.com/getsentry/sentry-dart/blob/b728df439473e146f8721c7555cd54728a93b6cb/dart/lib/src/sentry_client.dart#L292

Is it possible for transaction being captured through he `captureEvent` method? Then it makes sense to also pass on the hint.

File: dart/lib/src/sentry_client.dart (hunk 3 above)
Comment (denrase) [reply dart/lib/src/sentry_client.dart:391]: Looked over at `sentry-javascript`, the initial PR implementing this feature always used `beforeSend` as reason. We should probably check if the server knows about `beforeSendTransaction`?

File: dart/lib/src/sentry_options.dart (hunk 1 above)
Comment (marandaneto) [reply dart/lib/src/sentry_options.dart:436]: Good point, we can check the JS/Python implementation, maybe its not needed.
Docs https://github.com/getsentry/sentry-docs/pull/5814

File: dart/lib/src/sentry_client.dart (hunk 3 above)
Comment (marandaneto) [reply dart/lib/src/sentry_client.dart:391]: Just confirmed, we should keep using `beforeSend` for now.

Comment (denrase) [reply dart/lib/src/sentry_client.dart:391]: thx! 🙇

File: dart/lib/src/sentry_options.dart (hunk 1 above)
Comment (denrase) [reply dart/lib/src/sentry_options.dart:434]: Think it does not hurt to at least pass it in the case wehre a potential transaction comes through the captureError method.

Comment (denrase) [reply dart/lib/src/sentry_options.dart:436]: Just checked other SDKs. Java actually does provide the `Hint` param in the `captureTransaction` method. So we should probably also add this. To the API. WDYT?

Comment (marandaneto) [reply dart/lib/src/sentry_options.dart:436]: No need, the `startTransaction` methods don't take a `hint`, so users won't be able to pass it anyway, adding everywhere would be more work, we can add it later if needed.

File: dart/lib/src/sentry_client.dart (hunk 2 above)
Comment (marandaneto) [reply dart/lib/src/sentry_client.dart:377]: See https://github.com/getsentry/sentry-dart/pull/1238#discussion_r1085651408

File: dart/test/sentry_client_test.dart (hunk 4 above)
Comment (denrase) [reply dart/test/sentry_client_test.dart]: I fixed those issues. There were 2 properties that were never initialized if we don't pass it on through ctor. I'm not sure if they are relevant for the transaction and i didn't check if there are other params that need to be passed as well.

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,12 @@
 
 ## Unreleased
 
+### Features
+
+- Support beforeSendTransaction ([#1238](https://github.com/getsentry/sentry-dart/pull/1238))
+  - Add `fingerprint` parameter to `SentryTransaction` constructor
+  - Add `modules` parameter to `SentryTransaction` constructor
```
Comment (marandaneto) [inline CHANGELOG.md]: Can we double-check if sentry-java or sentry-cocoa also sends or does not send those 2 values for transactions? I don't remember by heart, @romtsn or @philipphofmann can help.

Comment (denrase) [reply CHANGELOG.md]: I guess those are not part of transaction. In `sentry-java`, the event and transaction are both subclasses of a base event. They do not share [`fingerprint`](https://github.com/getsentry/sentry-java/blob/f6a135d1ef81f9cd70d72608e8beb8c2103f69bd/sentry/src/main/java/io/sentry/SentryEvent.java#L69) and `modules`. Those are only part of the event. I'll remove those again and also remove them from the test expectation.

**Comment by marandaneto** [comment]
@denrase follow up about `modules` and `fingerprint`, since we are on the v7 branch, we can move them from the base class to the event class specifically, what do you think? since this is about the beforeSend, we can do it in a follow-up PR.

**Comment by denrase** [comment]
@marandaneto Yeah lets do this separate, if theres even something to do. Those are part of the event base class already, you can just never use them in the transaction subclass, but i guess they are not needed there.

**Pull request merged**